Added -whiteconnections=<n> option

Krellan opened the pull request:
This allows whitelisted peers a higher connection limit,
so they can exceed the normal -maxconnections limit,
useful for ensuring your local users and miners can always get in.

Krellan committed:
Added -whiteconnections=<n> option

This sets aside a number of connection slots for whitelisted peers,
useful for ensuring your local users and miners can always get in,
even if your limit on inbound connections has already been reached.

fanquake commented on src/net.h:
```diff
@@ -123,7 +123,17 @@ extern bool fListen;
 extern uint64_t nLocalServices;
 extern uint64_t nLocalHostNonce;
 extern CAddrMan addrman;
+
```
Could you update this to use our doxygen comment format.

laanwj replied:
It's not easy to integrate a 'floating' comment like this into doxygen.
One way would be to group nMaxConnections and nWhiteConnections using `@{` `@}` and add it as doxygen comment of the group.

Krellan commented:
Thanks, will update it.

21E14 commented:
utAck. Is the pending update keeping the merge up?

Krellan commented:
Yikes!  Forgot about this.  Will update ASAP.

Krellan commented:
There, got it merged with latest master, and made the requested documentation cleanup changes!

The Travis build still fails for the "no wallet" case, but the same thing happens right now for the latest upstream master....

Krellan commented:
No code change, just rebased with latest master to pick up the Travis build fix.

21E14 commented:
Thanks!

laanwj commented on src/init.cpp:
```diff
@@ -668,7 +668,29 @@ bool AppInit2(boost::thread_group& threadGroup)
+        // User not using whitelist feature.
+        // Silently disable connection reservation,
+        // for the same reason as above.
+        nWhiteConnections = 0;
```
What if the user is using `whitebind` instead?

Krellan commented:
Good to point out!  I will change it so that the "User is using whitelist feature" test matches on either -whitelist or -whitebind options.

Krellan commented:
There, updated it as described (both `whitelist` and `whitebind` will work to trigger the nMaxConnections fixup).  Also rebased against latest master.

laanwj commented on src/init.cpp:
```diff
@@ -668,7 +668,29 @@ bool AppInit2(boost::thread_group& threadGroup)
+    nWhiteConnections = std::max(nUserWhiteConnections, 0);
+
+    if ((mapArgs.count("-whitelist")) || (mapArgs.count("-whitebind"))) {
+        if (!(mapArgs.count("-maxconnections"))) {
```
This inner `if()` seems a bit overcomplicated. I think this addition should either be always done (when whitelisting is used), or never at all. To me it doesn't hold up to the principle of least surprise to do this only to the default `-maxconnections` but not when the user provides one. Too much parameter interaction...

Krellan commented:
I thought about that as well, during initial implementation thoughts over IRC.  Talked about it with a few others.  Tried to avoid surprise and unintended behavior as best as possible, and this was a compromise.

If the increase was done always, then it would be surprising to the user, because it would look like we were blatantly disobeying their given `maxconnections` limit.

If user doesn't provide `maxconnections` at all, then the thinking was that they wouldn't care as much, therefore we could get away with doing the increase.  We definitely do want to do the increase, as mentioned in the comments, to avoid reducing connection capacity of the Bitcoin network.

So, there's thinking that went into both sides of this decision.

21E14 commented:
Changing from utAck to tAck.

21E14 commented:
I am not sure what's been keeping the PR open?

Krellan commented:
I'm not sure either.  I'm hoping there is no issue preventing this from being merged.

Krellan commented:
Trivial change to keep up with latest upstream.

sipa commented:
@Krellan There are several unaddressed comments?

Krellan commented:
@sipa Which comments? Thought I addressed all concerns.

sipa commented:
Sorry, I missed that you answered in the PR rather than on the commits.

sipa commented:
The "increase the maxconnections if not specified, don't if it is" is fine, I think, but I wonder if the -whiteconnections default = 4 is needed. An existing user of one of the white* features won't expect a change in behaviour.

I wonder if this isn't a concise way to specify it:
* -whiteconnections defaults to 0, no matter what
* -maxconnections defaults to 125 + [-whiteconnections] (or whatever the FD limit permits)

Seems easier to explain, and wouldn't violate the least surprise principle either?

Krellan commented:
Good point.  I will make it default to 0 instead of 4, so that the default behavior is to change nothing at all.

Krellan commented:
There, as requested, changed the default from 4 to 0.  Made no other code changes, just these constants.  Also rebased to latest upstream.

So, the behavior is now unchanged, by default.  If the user wants the benefit of reserved slots for their whitelisted incoming connections, they will have to explicitly specify how many slots they want to reserve.

laanwj commented on src/init.cpp:
```diff
@@ -720,6 +710,48 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+    if (nFD < MIN_CORE_FILEDESCRIPTORS)
+        return InitError(_("Not enough file descriptors available."));
+    if (nFD - MIN_CORE_FILEDESCRIPTORS < nMaxConnections)
+        nMaxConnections = nFD - MIN_CORE_FILEDESCRIPTORS;
```
Slightly preferred for readability:
```c++
nMaxConnections = std::min(nFD - MIN_CORE_FILEDESCRIPTORS, nMaxConnections);
```

laanwj commented:
utACK apart from nit (but needs rebase)

Krellan commented:
Rebased.

Krellan commented:
Addressed the nit, good catch.

Also found another place where the default was still 4, changed to 0.

sipa commented:
Untested ACK
